src/test/java/io/confluent/connect/elasticsearch/bulk/BulkProcessorTest.java

```diff
@@ -0,0 +1,270 @@
+    }
+  }
+
+  @Test
+  public void batchingAndLingering() throws InterruptedException, ExecutionException {
+    final Client client = new Client();
+
+    final int maxBufferedRecords = 100;
+    final int maxInFlightBatches = 5;
+    final int batchSize = 5;
+    final int lingerMs = 5;
+    final int maxRetries = 0;
+    final int retryBackoffMs = 0;
+
+    final BulkProcessor<Integer, ?> bulkProcessor = new BulkProcessor<>(
+        new SystemTime(),
+        client,
+        maxBufferedRecords,
+        maxInFlightBatches,
+        batchSize,
+        lingerMs,
+        maxRetries,
+        retryBackoffMs
+    );
+
+    final int addTimeoutMs = 10;
+    bulkProcessor.add(1, addTimeoutMs);
+    bulkProcessor.add(2, addTimeoutMs);
+    bulkProcessor.add(3, addTimeoutMs);
+    bulkProcessor.add(4, addTimeoutMs);
+    bulkProcessor.add(5, addTimeoutMs);
+    bulkProcessor.add(6, addTimeoutMs);
+    bulkProcessor.add(7, addTimeoutMs);
+    bulkProcessor.add(8, addTimeoutMs);
+    bulkProcessor.add(9, addTimeoutMs);
+    bulkProcessor.add(10, addTimeoutMs);
+    bulkProcessor.add(11, addTimeoutMs);
+    bulkProcessor.add(12, addTimeoutMs);
+
+    client.expect(Arrays.asList(1, 2, 3, 4, 5), BulkResponse.success());
+    client.expect(Arrays.asList(6, 7, 8, 9, 10), BulkResponse.success());
+    client.expect(Arrays.asList(11, 12), BulkResponse.success()); // batch not full, but upon linger timeout
+    assertTrue(bulkProcessor.submitBatchWhenReady().get().succeeded);
+    assertTrue(bulkProcessor.submitBatchWhenReady().get().succeeded);
+    assertTrue(bulkProcessor.submitBatchWhenReady().get().succeeded);
+    assertTrue(client.expectationsMet());
+  }
+
+  @Test
+  public void flushing() {
+    final Client client = new Client();
+
+    final int maxBufferedRecords = 100;
+    final int maxInFlightBatches = 5;
+    final int batchSize = 5;
+    final int lingerMs = 100000; // super high on purpose to make sure flush is what's causing the request
+    final int maxRetries = 0;
+    final int retryBackoffMs = 0;
+
+    final BulkProcessor<Integer, ?> bulkProcessor = new BulkProcessor<>(
+        new SystemTime(),
+        client,
+        maxBufferedRecords,
+        maxInFlightBatches,
+        batchSize,
+        lingerMs,
+        maxRetries,
+        retryBackoffMs
+    );
+
+    client.expect(Arrays.asList(1, 2, 3), BulkResponse.success());
+
+    bulkProcessor.start();
+
+    final int addTimeoutMs = 10;
+    bulkProcessor.add(1, addTimeoutMs);
+    bulkProcessor.add(2, addTimeoutMs);
+    bulkProcessor.add(3, addTimeoutMs);
+
+    assertFalse(client.expectationsMet());
+
+    final int flushTimeoutMs = 10;
+    bulkProcessor.flush(flushTimeoutMs);
+
+    assertTrue(client.expectationsMet());
+  }
+
+  @Test
+  public void addBlocksWhenBufferFull() {
+    final Client client = new Client();
+
+    final int maxBufferedRecords = 1;
+    final int maxInFlightBatches = 1;
+    final int batchSize = 1;
+    final int lingerMs = 10;
```

Comment:
I'd probably make this very conservatively high any time you want lingering to have basically no effect. Sometimes jenkins gets surprisingly slow, especially when running multiple jobs at once.